src/main/kotlin/org/wfanet/measurement/reporting/service/api/v1alpha/SetOperationCompiler.kt

```diff
@@ -0,0 +1,509 @@
+// Copyright 2022 The Cross-Media Measurement Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package org.wfanet.measurement.reporting.service.api.v1alpha
+
+import kotlin.math.pow
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import org.wfanet.measurement.reporting.v1alpha.Metric.NamedSetOperation
+import org.wfanet.measurement.reporting.v1alpha.Metric.SetOperation
+
+/**
+ * A primitive region of a Venn diagram is the intersection of a set of reporting sets, and it is
+ * represented by a bit representation of an integer. Only the reporting sets with IDs equal to the
+ * bit positions of set bits constitute the primitive region. Ex: Given a Venn Diagram of 3
+ * reporting sets (rs0, rs1, rs2), a primitive region with an integer value equal to 3 has the bit
+ * representation b’011’. This means this primitive region is only covered by the intersection of
+ * rs0 and rs1 and not covered by rs2 (the order of the bit positions is from right to left). In
+ * other words, rs0 INTERSECT rs1 INTERSECT COMPLEMENT(rs2). Note that primitive regions are
+ * disjoint.
+ */
+private typealias PrimitiveRegion = ULong
+
+/**
+ * A union set is the union of a set of reporting sets, and it is represented by a bit
+ * representation of an integer. Only the reporting sets with IDs equal to the bit positions of set
+ * bits constitute the union set. Given a Venn Diagram of 3 reporting sets (rs0, rs1, rs2), a
+ * union-set with an integer value equal to 3 has the bit representation b’011’. This means this
+ * union-set is only covered by the union of rs0 and rs1 (the order of the bit positions is from
+ * right to left).
+ */
+private typealias UnionSet = ULong
+
+private typealias NumberReportingSets = Int
+
+/** A mapping from a [UnionSet] to its coefficient in the Venn diagram region decomposition. */
+private typealias UnionSetCoefficientMap = Map<UnionSet, Int>
+
+/**
+ * A mapping for cardinality computation from a [PrimitiveRegion] to its decomposition in terms of
+ * union-sets represented by [UnionSetCoefficientMap]. Take a case of 3 reporting sets (rs0, rs1,
+ * rs2) as an example. A primitive region with its value equal to 3 (b'011') is the intersection of
+ * rs0 and rs1. The decomposition of the cardinality of the region =
```

Comment:
Following the lines 28~30, should it be the intersection of rs0, and rs1, and COMPLEMENT(rs2)?